### PR TITLE
testing out typed dataframes

### PR DESCRIPTION
**Pull request by pepegar**
# Description

Here's my attempt at a more typed core for Bellman.  I'm creating this PR as more documentation and conversation space than something I'd like to merge to `master`.

**TLDR;**: Our hypothesis in which, by parsing strings to actual values would improve time, didn't work.

I think using using a more typed approach (either with Frameless, or plain Datasets) makes sense for ETLs, if someone is moving data from point to point and performing transformations inside.  On the other hand if what you (we in this case) is to create a general purpose query engine, and the shape of data can change on every expression that the user adds to the query, we need to use the most genera API possible, `DataFrames` in our case.  Otherwise, the performance penalty described in **`Deserialization`** is just too high.

In the file `RdfValuesSpec` you can see the two simple tests I wanted to perform.  On one hand I'm just calling a function implemented for untyped columns, and then the sample functino implemented for `RdfValues`. 

The first one is **noticeably** faster, and has far less heap & metaspace fingerprint.

# Conclusion

I think we shouldn't go for this approach for now.  It won't solve the issue as we expected.

Some takeaways of this spike:

## 👍🏻 Types

our functions could be typed when using this approach. See `def strafter` in the diff.  It looks much better than working with Spark column ops, and is less error prone.

## 👎🏻 Deserialization

If we want to use the typed APIs of either `frameless.TypedDataset`, or `spark.sql.DataSet`, we need our functions to operate on values on the heap.  In order to do that, these operations **must** run outside of catalyst, and Catalyst needs to convert from Tungsten representation to JVM and back again.

I expected typed functions to do the opposite, be serialized to catalyst operations instead.

### CODE REVIEW

**Comment by JNKHunter** [comment]
This is all part of fail fast! Good analysis and learnings.

**Comment by patofp** [comment]
> # Description
> Here's my attempt at a more typed core for Bellman. I'm creating this PR as more documentation and conversation space than something I'd like to merge to `master`.
> 
> **TLDR;**: Our hypothesis in which, by parsing strings to actual values would improve time, didn't work.
> 
> I think using using a more typed approach (either with Frameless, or plain Datasets) makes sense for ETLs, if someone is moving data from point to point and performing transformations inside. On the other hand if what you (we in this case) is to create a general purpose query engine, and the shape of data can change on every expression that the user adds to the query, we need to use the most genera API possible, `DataFrames` in our case. Otherwise, the performance penalty described in **`Deserialization`** is just too high.
> 
> In the file `RdfValuesSpec` you can see the two simple tests I wanted to perform. On one hand I'm just calling a function implemented for untyped columns, and then the sample functino implemented for `RdfValues`.
> 
> The first one is **noticeably** faster, and has far less heap & metaspace fingerprint.
> 
> # Conclusion
> I think we shouldn't go for this approach for now. It won't solve the issue as we expected.
> 
> Some takeaways of this spike:
> 
> ## 👍🏻 Types
> our functions could be typed when using this approach. See `def strafter` in the diff. It looks much better than working with Spark column ops, and is less error prone.
> 
> ## 👎🏻 Deserialization
> If we want to use the typed APIs of either `frameless.TypedDataset`, or `spark.sql.DataSet`, we need our functions to operate on values on the heap. In order to do that, these operations **must** run outside of catalyst, and Catalyst needs to convert from Tungsten representation to JVM and back again.
> 
> I expected typed functions to do the opposite, be serialized to catalyst operations instead.
